File: FindNthDigit/FindNthDigit.cpp
class Solution {
public:
    int findNthDigit(int n) {
        if (n<0)
            return -1;
        if (n<10)
            return n;
        int digits = 1;
        long number = 9;
        while  (n>digits*number)
        {
            n-=digits*number;
            digits++;
            number*=10;
        }
        n-=1; // 计算时从1开始，但实际是从0开始，所以要减1
        int  a = n / digits;
        int  idx = n%digits;
        int num = int (pow(10, digits-1)) + a;
        int idxFromRight = digits-idx;
        for (int i=1;i<idxFromRight;i++)
        {
            num/=10;
        }
        return num%10;
        // string s = to_string(num);
        // return s[idx]-'0';
    }
};

File: FindNthDigit/FindNthDigit.py
class Solution:
    def findNthDigit(self, n: int) -> int:
        if n<0:
            return -1
        if n<10:
            return n
        digits = 1
        numbers = 9
        while n>digits*numbers:
            n-=digits*numbers
            digits+=1
            numbers*=10
        n-=1
        a = n//digits
        b = n%digits
        num = int(pow(10,digits-1) + a)
        idx = digits-b
        for i in range(1, idx):
            num//=10
        return num%10
        # s = str(num)
        # res = int(s[b])
        # return res
